contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1024 @@
+   * - One of the following must be true:
+   *    - the operation is a mint (which should ONLY occur when BridgedPolygonNORI is being wrapped via `_depositFor`)
+   *    - the operation is a burn, which only happens during revocation and withdrawal:
+   *      - if the operation is a revocation, that permission is enforced by the TOKEN_REVOKER_ROLE
+   *      - if the operation is a withdrawal the burn amount must be <= the sender's claimable balance
+   *    - the operation is a transfer and _all_ of the following must be true:
+   *      - the operator is operating on their own balance (enforced in the inherited contract)
+   *      - the operator has sufficient balance to transfer (enforced in the inherited contract)
+   */
+  function _beforeTokenTransfer(
+    address operator,
+    address from,
+    address to,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    bytes memory data
+  ) internal override(ERC1155SupplyUpgradeable, ERC1155PausableUpgradeable) {
+    bool isBurning = to == address(0);
+    bool isWithdrawing = isBurning && from == operator;
+
+    if (isBurning) {
+      for (uint256 i = 0; i < ids.length; i++) {
+        uint256 id = ids[i];
+        Schedule storage schedule = _scheduleIdToScheduleStruct[id];
+        schedule.releasedAmountFloor = _releasedBalanceOfSingleSchedule(id);
+      }
+    }
+    if (isWithdrawing) {
+      for (uint256 i = 0; i < ids.length; i++) {
+        uint256 id = ids[i];
+        if (amounts[i] > claimableBalanceForScheduleForAccount(id, from)) {
+          revert InsufficientClaimableBalance({account: from, scheduleId: id});
+        }
+      }
+    }
+    return super._beforeTokenTransfer(operator, from, to, ids, amounts, data);
+  }
+
+  /**
+   * Linearly released balance for a single schedule at the current block timestamp, ignoring any
+   * released amount floor that has been set for the schedule.
+   */
+  /* solhint-disable not-rely-on-time, this is time-dependent */
+  function _linearReleaseAmountAvailable(uint256 scheduleId)
+    internal
+    view
+    returns (uint256)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    uint256 linearAmountAvailable;
+    if (block.timestamp >= schedule.endTime) {
+      linearAmountAvailable = totalSupply(scheduleId);
+    } else {
+      uint256 rampTotalTime = schedule.endTime - schedule.startTime;
+      linearAmountAvailable = block.timestamp < schedule.startTime
+        ? 0
+        : (totalSupply(scheduleId) * (block.timestamp - schedule.startTime)) /
+          rampTotalTime;
+    }
+    return linearAmountAvailable;
+  }
+
+  /* solhint-enable not-rely-on-time */
+
+  /**
+   * @dev Calculates the number of tokens that should be revoked from a given token holder and schedule based on their
+   * proportion of ownership of that schedule's tokens and the total number of tokens being revoked.
+   */
+  function _quantityToRevokePerTokenHolder(
+    uint256 totalQuantityToRevoke,
```

Comment:
Per comment above, I think this could be refactored to be clearer by making this function return percent ownership and leave the application of that to the revocation amount to the calling function.

Reply:
again, precision problems